File: server/Allay/src/main/java/cn/huohuas001/huhobot/allay/utils/CollectionUtils.java
package cn.huohuas001.huhobot.allay.utils;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public final class CollectionUtils {
    /**
     * 将一个Set集合转换为List，然后分割成多个子List。
     *
     * @param set  要分割的Set集合。
     * @param size 每个子List的最大容量。
     * @param <T>  Set集合中元素的类型。
     * @return 分割后的List列表。
     */
    public static <T> List<List<T>> chunkSet(Set<T> set, int size) {
        // 将Set转换为List
        var list = new ArrayList<>(set);
        // 使用chunkList方法进行分片
        return chunkList(list, size);
    }

    private static <T> List<List<T>> chunkList(List<T> list, int size) {
        var chunks = new ArrayList<List<T>>();
        for (int i = 0; i < list.size(); i += size) {
            var chunk = list.subList(i, Math.min(i + size, list.size()));
            chunks.add(chunk);
        }
        return chunks;
    }

    public static List<String> searchInSet(Set<String> set, String keyword) {
        return set.stream()
                .filter(s -> s.contains(keyword))
                .collect(Collectors.toList());
    }
}

File: server/Architectury/common/src/main/kotlin/cn/huohuas001/huhobot/mod/tools/CollectionUtils.kt
package cn.huohuas001.huhobot.mod.tools

import net.minecraft.server.players.UserWhiteList

object CollectionUtils {
    /**
     * 将一个Set集合转换为List，然后分割成多个子List。
     *
     * @param set  要分割的Set集合。
     * @param size 每个子List的最大容量。
     * @return 分割后的List列表。
     */
    fun <T> chunkSet(set: Set<T>, size: Int): List<List<T>> {
        val list = ArrayList<T>(set)
        return chunkList(list, size)
    }

    fun chuckUserList(list: UserWhiteList, size: Int): List<List<String>>{
        val userList = list.userList
        return chunkList(userList.toList(), size)
    }

    private fun <T> chunkList(list: List<T>, size: Int): List<List<T>> {
        val chunks = ArrayList<List<T>>()
        for (i in 0 until list.size step size) {
            val chunk = list.subList(i, (i + size).coerceAtMost(list.size))
            chunks.add(chunk)
        }
        return chunks
    }

    /**
     * 在Set中搜索包含指定关键字的元素。
     *
     * @param set     要搜索的Set集合。
     * @param keyword 关键字。
     * @return 包含关键字的元素列表。
     */
    fun searchInSet(set: Set<String>, keyword: String): List<String> {
        return set.filter { it.contains(keyword) }.toList()
    }

    fun searchInUserWhiteList(list: UserWhiteList, keyword: String): List<String> {
        val userList = list.userList
        return searchInSet(userList.toSet(), keyword)
    }
}
